File: singleylinkedlist/sll.cpp
//Copyright Asaad 2018 - ∞

#ifndef SLL_H
#define SLL_H

#include <iostream>

template<typename T>
class SLL {

   public:
   
      struct Node {
         T Data;
         Node* Next;
      };
      
      void Insert(const T Data);
      void Update(const T Data , const T UpdatedData);
      void Remove(const T Data);
      void Print() {
         Curr = Head;
         while(Curr != nullptr) {
            std::cout << Curr->Data << " ";
            Curr = Curr->Next;
         }
      }
      
      Node* Find(const T Data) {
         Curr = Head;
         while(Curr->Next != nullptr && Curr->Data != Data)
            Curr = Curr->Next;
         if(Curr->Data == Data) {
            std::cout << "Found Data";
            return Curr;
         } else {
            std::cout << "Could not Find Data in the list \n";
            return nullptr;
         }
      }
      
   private:
   
      Node* Head = nullptr;
      Node* Curr = nullptr;     
};

#endif

template<typename T> void SLL<T>::Insert(const T Data) {
   Node* newNode = new Node();
   newNode->Data = Data;
   newNode->Next = nullptr;
   
   if(Head == nullptr)
      Head = newNode;
   else {
      Curr = Head;
      while(Curr->Next != nullptr)
         Curr = Curr->Next;
      Curr->Next = newNode;
   }
}

template<typename T> void SLL<T>::Update(const T Data , const T UpdatedData) {
   Node* SearchNode = Find(Data);
   if(!SearchNode)
      std::cout << "Could not update Value";
   else 
      SearchNode->Data = UpdatedData;
}

template<typename T> void SLL<T>::Remove(const T Data) {
   Curr = Head;
   Node* Temp = nullptr;
   while(Curr->Next != nullptr && Curr->Data != Data) {
      Temp = Curr;
      Curr = Curr->Next;
   } if(Data == Head->Data) {
      Head = Curr->Next;
      delete Curr;
   } else if(Curr->Data == Data) {
      Temp->Next = Curr->Next;
      delete Curr;
   }
}

int main() {
   SLL<int> Test;
   Test.Insert(1);
   Test.Insert(2);
   Test.Insert(3);
   Test.Insert(4);
   Test.Remove(1);
   Test.Update(1,333);
   Test.Print();
}

File: singleylinkedlist/sll.js
//Copyright Asaad 2018 - ∞

class Node {
   constructor(Data,Next) {
      this.Data = Data;
      this.Next = Next;
   }
}

class SLL {
   constructor() {
      this.Head = null;
      this.Curr = null;
   }
   
   Insert(Data) {
      this.newNode = new Node(Data,null);
      if(this.Head == null)
         this.Head = this.newNode;
      else {
         this.Curr = this.Head;
         while(this.Curr.Next != null) {
            this.Curr = this.Curr.Next;
         } this.Curr.Next = this.newNode;
      }
   }
   
   Find(Data) {
      this.Curr = this.Head;
      while(this.Curr.Next != null && this.Curr.Data != Data)
         this.Curr = this.Curr.Next;
      if(this.Curr.Data === Data) 
         return this.Curr;
      else
         return null;
    }
    
    Update(Data,UpdatedData) {
       this.SearchNode = this.Find(Data);
       if(!this.SearchNode)
          console.log("Failed to update the data");
       this.SearchNode.Data = UpdatedData;
    }
   
   Remove(Data) {
      this.Curr = this.Head;
      this.Temp = new Node(null,null);
      while(this.Curr.Next != null && this.Curr.Data != Data) {
         this.Temp = this.Curr;
         this.Curr = this.Curr.Next;
      } if(this.Data === this.Head.Data) {
         this.Head = this.Curr.Next;
         this.Curr = null;
      } else if(this.Curr.Data === Data) {
         this.Temp.Next = this.Curr.Next;
         this.Curr = null;
      }
   }
   
   Print() {
      this.Curr = this.Head;
      while(this.Curr != null) {
         console.log(this.Curr.Data + " ");
         this.Curr = this.Curr.Next;
      }
   }
}

var Test = new SLL();

Test.Insert(1);
Test.Insert(2);
Test.Insert(3);

Test.Update(1,111);
Test.Remove(2);

Test.Print();
